trace_replay/trace_replay.cc

```diff
@@ -508,132 +504,73 @@ Status Replayer::Replay() {
-      iter_payload.cf_id = 0;
-      if (trace_file_version_ < 2) {
-        DecodeCFAndKey(trace.payload, &iter_payload.cf_id,
-                       &iter_payload.iter_key);
-      } else {
-        TracerHelper::DecodeIterPayload(&trace, &iter_payload);
-      }
-      if (iter_payload.cf_id > 0 &&
-          cf_map_.find(iter_payload.cf_id) == cf_map_.end()) {
-        return Status::Corruption("Invalid Column Family ID.");
-      }
+Status ReplayerImpl::Step() {
+  Status s =
+      StepWork(std::chrono::system_clock::time_point::min(), 0, 1, nullptr);
+  if (s.IsIncomplete() || s.IsNotSupported()) {
+    Reset().PermitUncheckedError();
+  }
+  return s;
+}
 
-      if (iter_payload.cf_id == 0) {
-        single_iter = db_->NewIterator(roptions);
-      } else {
-        single_iter = db_->NewIterator(roptions, cf_map_[iter_payload.cf_id]);
-      }
-      single_iter->Seek(iter_payload.iter_key);
-      ops++;
-      delete single_iter;
-    } else if (trace.type == kTraceIteratorSeekForPrev) {
-      // Currently, we only support to call SeekForPrev. The Next() and Prev()
-      // is not supported.
-      IterPayload iter_payload;
-      iter_payload.cf_id = 0;
-      if (trace_file_version_ < 2) {
-        DecodeCFAndKey(trace.payload, &iter_payload.cf_id,
-                       &iter_payload.iter_key);
-      } else {
-        TracerHelper::DecodeIterPayload(&trace, &iter_payload);
-      }
-      if (iter_payload.cf_id > 0 &&
-          cf_map_.find(iter_payload.cf_id) == cf_map_.end()) {
-        return Status::Corruption("Invalid Column Family ID.");
-      }
+Status ReplayerImpl::Replay(ReplayOptions options) {
+  if (options.fast_forward <= 0.0) {
+    return Status::InvalidArgument("Wrong fast forward speed!");
+  }
 
-      if (iter_payload.cf_id == 0) {
-        single_iter = db_->NewIterator(roptions);
-      } else {
-        single_iter = db_->NewIterator(roptions, cf_map_[iter_payload.cf_id]);
-      }
-      single_iter->SeekForPrev(iter_payload.iter_key);
-      ops++;
-      delete single_iter;
-    } else if (trace.type == kTraceMultiGet) {
-      MultiGetPayload multiget_payload;
-      assert(trace_file_version_ >= 2);
-      TracerHelper::DecodeMultiGetPayload(&trace, &multiget_payload);
-      std::vector<ColumnFamilyHandle*> v_cfd;
-      std::vector<Slice> keys;
-      assert(multiget_payload.cf_ids.size() ==
-             multiget_payload.multiget_keys.size());
-      for (size_t i = 0; i < multiget_payload.cf_ids.size(); i++) {
-        assert(i < multiget_payload.cf_ids.size() &&
-               i < multiget_payload.multiget_keys.size());
-        if (cf_map_.find(multiget_payload.cf_ids[i]) == cf_map_.end()) {
-          return Status::Corruption("Invalid Column Family ID.");
-        }
-        v_cfd.push_back(cf_map_[multiget_payload.cf_ids[i]]);
-        keys.push_back(Slice(multiget_payload.multiget_keys[i]));
+  {
+    std::lock_guard<std::mutex> guard(mutex_);
+    if (!prepared_) {
```

Comment:
If it is not Prepared, do you think it is better to fail the replay, instead of prepare it internally? We may need an explicit prepare

Reply:
I prefer to have on-demand auto prepare, but allow the user to reset.